assertion/function/assertiontree/trusted_func.go

```diff
@@ -180,74 +181,110 @@ var requireComparators action = func(call *ast.CallExpr, startIndex int, pass *a
+		_greaterThanZero
+		_nil
+	)
 
-		v, ok := constant.Val(typeAndValue.Value).(int64)
-		if !ok {
-			continue
-		}
+	var actualExpr ast.Expr
+	var actualExprIndex int
+	expectedExprValue := _unknown
 
-		// Now, based on the semantics of the function, we can create artificial nonnil checks for
-		// the following cases.
-		switch funcName {
-		case "Equal", "Equalf": // len(s) == [positive_int]
-			if v > 0 {
-				return newNilBinaryExpr(sliceExpr, token.NEQ)
+	for argIndex, expr := range call.Args[startIndex : startIndex+2] {
+		switch expr := expr.(type) {
+		case *ast.CallExpr:
+			// Check if the expression is `len(<slice_expr>)`.
+			wrapperFunc, ok := expr.Fun.(*ast.Ident)
+			if !ok {
+				continue
 			}
-		case "NotEqual", "NotEqualf": // len(s) != [zero]
-			if v == 0 {
-				return newNilBinaryExpr(sliceExpr, token.NEQ)
+			if pass.TypesInfo.ObjectOf(wrapperFunc) != util.BuiltinLen || len(expr.Args) != 1 {
+				continue
 			}
-		// Note the check for `argIndex` in the following cases, we need to make sure the slice expr
-		// is at the correct position since these are inequality checks.
-		case "Greater", "Greaterf": // len(s) > [non_negative_int]
-			if argIndex == 0 && v >= 0 {
-				return newNilBinaryExpr(sliceExpr, token.NEQ)
+			// Check if `<slice_expr>` is of slice type.
+			sliceExpr, lenExpr := expr.Args[0], call.Args[startIndex+1-argIndex]
+			_, ok = pass.TypesInfo.TypeOf(sliceExpr).Underlying().(*types.Slice)
+			if !ok {
+				continue
 			}
-		case "GreaterOrEqual", "GreaterOrEqualf": // len(s) >= [positive_int]
-			if argIndex == 0 && v > 0 {
-				return newNilBinaryExpr(sliceExpr, token.NEQ)
+
+			// Then, we can treat the other argument as the length expression and check its
+			// compile-time value.
+			typeAndValue, ok := pass.TypesInfo.Types[lenExpr]
+			if !ok {
+				continue
 			}
-		case "Less", "Lessf": // [non_negative_int] < len(s)
-			if argIndex == 1 && v >= 0 {
-				return newNilBinaryExpr(sliceExpr, token.NEQ)
+
+			v, ok := constant.Val(typeAndValue.Value).(int64)
+			if !ok {
+				continue
+			}
+
+			actualExpr = sliceExpr
+			actualExprIndex = argIndex
+			if v == 0 {
+				expectedExprValue = _zero
+			} else if v > 0 {
+				expectedExprValue = _greaterThanZero
 			}
-		case "LessOrEqual", "LessOrEqualf": // [positive_int] <= len(s)
-			if argIndex == 1 && v > 0 {
-				return newNilBinaryExpr(sliceExpr, token.NEQ)
+
+		case *ast.Ident:
+			// Check if the expected expression is `nil`.
+			if expr.Name == "nil" {
+				actualExpr = call.Args[startIndex+1-argIndex]
+				actualExprIndex = argIndex
+				expectedExprValue = _nil
 			}
 		}
 	}
 
+	// likely represents a case that we don't support.
+	if expectedExprValue == _unknown {
+		return nil
+	}
+
+	// Now, based on the semantics of the function, we can create artificial nonnil checks for
+	// the following cases.
+	switch funcName {
+	case "Equal", "Equalf": // len(s) == [positive_int], expr == nil
+		if expectedExprValue == _greaterThanZero {
```

Comment:
nit: Maybe add a comment above the switch re-stating the two cases we are supporting (length check + nil check), such that if the expected expr value is an int, the `actualExpr` is the slice expr inside the `len()` call.

I'm just thinking when we add more cases here in the future we might forget this fact and wonder what `actualExpr` is when `expectedExprValue == _greaterThanZero`.

Reply:
Good point. Added a comment.